### PR TITLE
Packages: add RubyGems support

### PR DESCRIPTION
**Pull request by olafurpg**
## Test plan

Steps to test the change locally:

- Create a "Ruby Dependencies" code host, it's OK to leave `dependencies` and `repository` empty 
- Visit the URL https://sourcegraph.test:3443/rubygems/actioncable@v7.0.0/-/blob/lib/action_cable.rb the package should get cloned automatically and the `action_cable.rb` file should load

### CODE REVIEW

File: sg.config.yaml [hunk 1]
```diff
@@ -154,7 +154,7 @@ commands:
       # TODO: This should be fixed
       export SOURCEGRAPH_LICENSE_GENERATION_KEY=$(cat ../dev-private/enterprise/dev/test-license-generation-key.pem)
       # If EXTSVC_CONFIG_FILE is *unset*, set a default.
-      export EXTSVC_CONFIG_FILE=${EXTSVC_CONFIG_FILE-'../dev-private/enterprise/dev/external-services-config.json'}
+      # export EXTSVC_CONFIG_FILE=${EXTSVC_CONFIG_FILE-'../dev-private/enterprise/dev/external-services-config.json'}
```
Comment (mrnugget) [inline sg.config.yaml:157]: Should this be here?

File: cmd/gitserver/server/vcs_syncer_ruby_packages.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+	if err != nil {
+		return errors.Wrap(err, "failed to create a temporary directory")
+	}
+	//defer os.RemoveAll(tmpDir.Name())
+
+	if err := unpack.Tar(r, tmpDir, opts); err != nil {
+		return errors.Wrapf(err, "failed to tar %s", tmpDir)
+	}
+
+	fmt.Printf("tmpDir=%s \n", tmpDir)
```
Comment (mrnugget) [inline cmd/gitserver/server/vcs_syncer_ruby_packages.go:100]: Debug print

File: cmd/gitserver/server/vcs_syncer_ruby_packages.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+		Filter: func(path string, file fs.FileInfo) bool {
+			return path == "data.tar.gz"
+		},
+	}
+
+	tmpDir, err := os.MkdirTemp("", "rubygems")
+	if err != nil {
+		return errors.Wrap(err, "failed to create a temporary directory")
+	}
+	//defer os.RemoveAll(tmpDir.Name())
```
Comment (mrnugget) [inline cmd/gitserver/server/vcs_syncer_ruby_packages.go:94]: This should probably stay in there, no?

File: internal/types/secret.go [hunk 4]
```diff
@@ -77,6 +77,8 @@ func (e *ExternalService) RedactedConfig(ctx context.Context) (string, error) {
 		}
 	case *schema.RustPackagesConnection:
 		// Nothing to redact
+	case *schema.RubyPackagesConnection:
+		es.redactString(c.Repository, "repository")
```
Comment (mrnugget) [inline internal/types/secret.go:81]: Why does this need to be redacted? Because of auth info in the URL?

**Review by mrnugget — COMMENTED** [review]
(Just some drive-by comments after seeing PR in #packages, don't block on me)

File: cmd/gitserver/server/vcs_syncer_ruby_packages.go (hunk 3 above)
Comment (olafurpg) [reply cmd/gitserver/server/vcs_syncer_ruby_packages.go]: Good catch! I commented it out to debug a download error and forgot to add it back.

File: cmd/gitserver/server/vcs_syncer_ruby_packages.go (hunk 2 above)
Comment (olafurpg) [reply cmd/gitserver/server/vcs_syncer_ruby_packages.go]: Removed

File: internal/types/secret.go (hunk 4 above)
Comment (olafurpg) [reply internal/types/secret.go:81]: Correct, see the Artifactory Ruby docs https://www.jfrog.com/confluence/display/JFROG/RubyGems+Repositories

> Use a source URL with embedded credentials, such as: gem sources -a http://user:password@host/...

CleanShot 2022-10-11 at 16 37 56

File: sg.config.yaml (hunk 1 above)
Comment (olafurpg) [reply sg.config.yaml]: It shouldn't! Removed.

**Review by olafurpg — COMMENTED** [review]
@mrnugget thank you for the review! Those were good catches

File: cmd/gitserver/server/vcs_syncer_ruby_packages.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+	repositoryURL := connection.Repository
+	if repositoryURL == "" {
+		repositoryURL = "https://rubygems.org/gems"
+	}
+	return &vcsPackagesSyncer{
+		logger:      log.Scoped("RubyPackagesSyncer", "sync Ruby packages"),
+		typ:         "ruby_packages",
+		scheme:      dependencies.RubyPackagesScheme,
+		placeholder: reposource.NewRubyVersionedPackage("shopify_api", "12.0.0"),
```
Comment (Strum355) [inline cmd/gitserver/server/vcs_syncer_ruby_packages.go:37]: should this be a different placeholder name? :P

File: cmd/gitserver/server/vcs_syncer_ruby_packages.go [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+		return errors.Wrap(err, "failed to create a temporary directory")
+	}
+	defer os.RemoveAll(tmpDir)
+
+	if err := unpack.Tar(r, tmpDir, opts); err != nil {
+		return errors.Wrapf(err, "failed to tar %s", tmpDir)
+	}
+
+	dataTarGzPath := filepath.Join(tmpDir, "data.tar.gz")
+	dataTarGzBytes, err := os.ReadFile(dataTarGzPath)
```
Comment (Strum355) [inline cmd/gitserver/server/vcs_syncer_ruby_packages.go:101]: I would likely recommend to use `os.Open` here to get an `io.ReadCloser` here. `unpack.Tgz` seems to use `io.ReadFull` on the `io.Reader` down below, so this would reduce the amount of copies of the file we have in-memory by one

**Review by Strum355 — APPROVED** [review]
LGTM besides the small points

File: client/web/src/components/externalServices/externalServices.tsx [hunk 7]
```diff
@@ -1399,6 +1401,39 @@ const RUST_PACKAGES = {
...
+    defaultDisplayName: 'Ruby Dependencies',
+    defaultConfig: `{
+  "repository": "https://rubygems.org/gems",
+  "dependencies": []
+}`,
+    instructions: (
+        <div>
+            <ol>
+                <li>
+                    The URL https://rubygems.org/gems is used if the field <Code>"repository"</Code> is empty.
```
Comment (varungandhi-src) [inline client/web/src/components/externalServices/externalServices.tsx:1418]: nit: Mention the `gem_name@version` syntax here in context.

**Comment by varungandhi-src** [comment]
Can you add wiring for auto-indexing and authorization in this PR too? Or are those meant for follow-up PRs?

File: cmd/gitserver/server/vcs_syncer_ruby_packages.go [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+func (rubyDependencySource) ParsePackageFromName(name reposource.PackageName) (reposource.Package, error) {
+	return reposource.ParseRubyPackageFromName(name)
+
+}
+func (rubyDependencySource) ParsePackageFromRepoName(repoName api.RepoName) (reposource.Package, error) {
+	return reposource.ParseRubyPackageFromRepoName(repoName)
+}
+
+func (s *rubyDependencySource) Download(ctx context.Context, dir string, dep reposource.VersionedPackage) error {
+	packageURL := fmt.Sprintf("%s/%s-%s.gem", s.repositoryURL, dep.PackageSyntax(), dep.PackageVersion())
```
Comment (varungandhi-src) [inline cmd/gitserver/server/vcs_syncer_ruby_packages.go:66]: Instead of relying on implementation details here, I think we should stick to the official API. Yes, it costs 2 requests, but latency is not really the slow part here, and the JSON blob is small.

1. The root `repositoryURL` should be `rubygems.org`, not `rubygems.org/gems`. That way we only need to add a suffix, not cut the suffix first.
2. Use this API https://guides.rubygems.org/rubygems-org-api-v2/ and read the `gem_uri` key, and use that to download the gem.

File: cmd/gitserver/server/vcs_syncer_ruby_packages.go [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+	pkg, err := s.client.Get(ctx, packageURL)
+	if err != nil {
+		return errors.Wrapf(err, "error downloading RubyGem with URL '%s'", packageURL)
```
Comment (varungandhi-src) [inline cmd/gitserver/server/vcs_syncer_ruby_packages.go:70]: nit: This error should include the name and version, as the URL will not be controlled by us.

File: cmd/gitserver/server/vcs_syncer_ruby_packages.go [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+	}
+
+	if err = unpackRubyPackage(pkg, dir); err != nil {
+		return errors.Wrap(err, "failed to unzip ruby module")
```
Comment (varungandhi-src) [inline cmd/gitserver/server/vcs_syncer_ruby_packages.go:74]: nit: This error should probably include the name, version and URL for debuggability.

File: doc/admin/external_service/ruby.md [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+1. Press **Add repositories**.
+
+## Repository syncing
+
+There are two ways to sync Ruby dependency repositories.
+
+* **Indexing** (recommended): run [`scip-ruby`](https://sourcegraph.github.io/scip-ruby/) against your Ruby codebase and upload the generated index to Sourcegraph using the [src-cli](https://github.com/sourcegraph/src-cli) command `src code-intel upload`. This is usually setup to run in a CI pipeline. Sourcegraph automatically synchronizes Ruby dependency repositories based on the dependencies that are discovered by `scip-ruby`.
+* **Code host configuration**: manually list dependencies in the `"dependencies"` section of the [JSON configuration](#configuration) when creating the JVM dependency code host. This method can be useful to verify that the credentials are picked up correctly without having to upload an index.
+
+## Credentials
```
Comment (varungandhi-src) [inline doc/admin/external_service/ruby.md:26]: This section and the sections below need updating.

File: doc/integration/ruby.md [hunk 12]
```diff
@@ -0,0 +1,29 @@
+# Ruby dependencies integration with Sourcegraph
+
+You can use Sourcegraph with Ruby dependencies from any RubyGem repository, including rubygems.org or an internal Artifactory.
```
Comment (varungandhi-src) [inline doc/integration/ruby.md:3]:
```suggestion
You can use Sourcegraph with Ruby dependencies from any RubyGems repository, including rubygems.org or an internal Artifactory.
```

File: docker-images/syntax-highlighter/go.mod [hunk 13]
```diff
@@ -0,0 +1 @@
+module github.com/sourcegraph/sourcegraph/docker-images/syntax-highlighter
```
Comment (varungandhi-src) [inline docker-images/syntax-highlighter/go.mod:1]: Unrelated change snuck in here.

File: schema/ruby-packages.schema.json [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "$id": "ruby-packages.schema.json#",
+  "title": "RubyPackagesConnection",
+  "description": "Configuration for a connection to Ruby packages",
+  "allowComments": true,
+  "type": "object",
+  "additionalProperties": false,
+  "properties": {
+    "repository": {
+      "description": "The url at which the maven repository can be found.",
```
Comment (varungandhi-src) [inline schema/ruby-packages.schema.json:11]:
```suggestion
      "description": "The url at which the RubyGems repository can be found.",
```

File: schema/ruby-packages.schema.json [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+      "type": "string",
+      "default": ["https://rubygems.org/gems"],
+      "examples": ["https://rubygems.org/gems", "https://<server name>.jfrog.io/artifactory/api/gems/<repository key>"]
```
Comment (varungandhi-src) [inline schema/ruby-packages.schema.json:14]: These shouldn't have the `/gems` suffix.

File: schema/ruby-packages.schema.json [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+        "requestsPerHour": {
+          "description": "Requests per hour permitted. This is an average, calculated per second. Internally, the burst limit is set to 100, which implies that for a requests per hour limit as low as 1, users will continue to be able to send a maximum of 100 requests immediately, provided that the complexity cost of each request is 1.",
+          "type": "number",
+          "default": 3600,
+          "minimum": 0
+        }
+      },
+      "default": {
+        "enabled": true,
+        "requestsPerHour": 3600
```
Comment (varungandhi-src) [inline schema/ruby-packages.schema.json:36]: This 3600 per hour is out of sync with the 10 requests/s limit in the code itself.

File: cmd/gitserver/server/vcs_syncer_ruby_packages.go [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+	}
+
+	tmpDir, err := os.MkdirTemp("", "rubygems")
+	if err != nil {
+		return errors.Wrap(err, "failed to create a temporary directory")
+	}
+	defer os.RemoveAll(tmpDir)
+
+	if err := unpack.Tar(r, tmpDir, opts); err != nil {
+		return errors.Wrapf(err, "failed to tar %s", tmpDir)
```
Comment (varungandhi-src) [inline cmd/gitserver/server/vcs_syncer_ruby_packages.go:97]: nit: This error message is incorrect, the problem is we failed to untar the bytes passed in `pkg` and which are being read by the reader `r`.

File: doc/admin/external_service/index.md [hunk 18]
```diff
@@ -141,6 +141,7 @@ We recognize there are other code hosts including CVS, Azure Dev Ops, SVN, and m
   - [Go dependencies](go.md)
   - [npm dependencies](npm.md)
   - [Python dependencies](python.md)
+  - [Ruby dependencies](ruby.md)
```
Comment (varungandhi-src) [inline doc/admin/external_service/index.md:144]: Hmm, why is this list missing Rust...

File: doc/integration/index.md [hunk 19]
```diff
@@ -15,6 +15,7 @@ Sourcegraph integrates with your other tools to help you search, navigate, and r
   - [npm dependencies](npm.md)
   - [Python dependencies](python.md)
   - [Go dependencies](go.md)
+  - [Ruby dependencies](ruby.md)
```
Comment (varungandhi-src) [inline doc/integration/index.md:18]: Unrelated to this PR but this list is also missing Rust.

**Review by olafurpg — COMMENTED** [review]
@varungandhi-src Thank you for the review!

File: cmd/gitserver/server/vcs_syncer_ruby_packages.go (hunk 8 above)
Comment (olafurpg) [reply cmd/gitserver/server/vcs_syncer_ruby_packages.go]: One big difference between two URLs is that the tarball URL can be behind a CDN while the API URL includes metadata like download counts that can be expensive to compute. We got blocked by crates.io for using their API and got the recommendation to use the CDN instead (computing the raw download URL ourselves).

Comment (olafurpg) [reply cmd/gitserver/server/vcs_syncer_ruby_packages.go]: Ideally, we want to be able to download all packages from rubygems.org in the future, and it could make a big difference if we use only CDN-able URLs

File: schema/ruby-packages.schema.json (hunk 16 above)
Comment (olafurpg) [reply schema/ruby-packages.schema.json:36]: I don't think this default is used anywhere, it doesn't get generated in the Go code and it doesn't appear in autocompletions for the settings editor.

**Review by varungandhi-src — APPROVED** [review]
Can you add some snapshot tests similar to the ones we have for npm?

If it is possible to test with a custom Artifactory instance with a free trial or something that'd be good to do, because it'd give greater confidence that we didn't miss anything and that it will work for customers using Artifactory.

**Comment by varungandhi-src** [comment]
I was looking at this code again, and it looks we're almost entirely missing instrumentation, so we will not have visibility into whether some downloads calls are slow etc. I will add that in follow-up PRs.

**Comment by varungandhi-src** [comment]
Follow up PR to add a test case. https://github.com/sourcegraph/sourcegraph/pull/43218
